src/librustc/mir/interpret/mod.rs

```diff
@@ -199,23 +227,48 @@ pub fn specialized_encode_alloc_id<
 
 pub fn specialized_decode_alloc_id<
     'a, 'tcx,
-    D: Decoder,
-    CACHE: FnOnce(&mut D, AllocId),
+    D: TyDecoder<'a, 'tcx>,
+    GlobalCache: FnMut(&mut D) -> LockGuard<'_, FxHashMap<usize, (AllocId, bool)>>,
+    LocalCache: FnOnce(&mut D) -> &mut FxHashSet<AllocId>,
 >(
     decoder: &mut D,
     tcx: TyCtxt<'a, 'tcx, 'tcx>,
-    cache: CACHE,
+    mut global_cache: GlobalCache,
+    local_cache: LocalCache,
 ) -> Result<AllocId, D::Error> {
+    let pos = decoder.position();
     match AllocKind::decode(decoder)? {
+        AllocKind::AllocAtPos => {
+            let real_pos = decoder.read_usize()?;
+            decoder.with_position(real_pos, AllocId::decode)
+        },
         AllocKind::Alloc => {
-            let alloc_id = tcx.alloc_map.lock().reserve();
-            trace!("creating alloc id {:?}", alloc_id);
-            // insert early to allow recursive allocs
-            cache(decoder, alloc_id);
+            // Read the size of the allocation.
+            // Used to skip ahead if we don't need to decode this.
+            let alloc_size = usize::decode(decoder)?;
+
+            let (alloc_id, fully_loaded) = *global_cache(decoder).entry(pos).or_insert_with(|| {
+                // Create an id which is not fully loaded
+                (tcx.alloc_map.lock().reserve(), false)
+            });
+            if fully_loaded || !local_cache(decoder).insert(alloc_id) {
```

Comment:
What happens if one thread tstarts decoding, the next thread takes over the CPU, gets here for the same AllocId, skips over and tries to access the allocation? It'll ICE about uncached alloc or error with dangling pointer, right?

Reply:
In the `fully_loaded` case this isn't a problem since the `AllocId` has an `Allocation` assigned. For the `!local_cache(decoder).insert(alloc_id)` case, we know that some stack frame above us will assign an `AllocId` before the result will be used. Since `local_cache` is thread local another thread won't see the value inserted here. It may instead decode the same allocation in parallel.

Reply:
Ah, neat. Please make this explanation a comment on that if statement